### ISSUE TITLE
draftfile.php / pluginfile.php 404 for files imported by repository_omeka — pathnamehash mismatch, not preview-generation

### ISSUE THREAD

**Issue by erseco** [ateeducacion/moodle-playground#92]
## Summary

Files imported through `repository_omeka` (and possibly other URL-downloading repositories) end up with a `mdl_files` row whose **`pathnamehash` does NOT match what Moodle's `draftfile.php` / `pluginfile.php` compute from the URL**. Every fetch of the file's URL (with or without `?preview=…`, with or without `.jpg`, with or without `forcedownload=1`) returns **404 with `filenotfound`** even though the file is fully ingested and visible in the file-picker grid.

This is the real cause of the broken thumbnails. PR #91 ("retry without `preview=`") is **not** the fix for this — it's a different, defensive safeguard for GD failures. Keep PR #91; this issue tracks the real underlying bug.

## Repro

1. Visit <https://moodle-playground.com/?blueprint-url=https://raw.githubusercontent.com/ateeducacion/moodle-repository_omeka/refs/heads/main/blueprint.json&debug=1>
2. Profile → Private files → **Add…** → Omeka Sandbox → Sherlockiana → pick any image → **Select this file**
3. The file appears in the picker grid with title + correct dimensions but a broken-thumbnail icon.
4. Open DevTools → Network: every `draftfile.php/.../<filename>` request returns **404**.

### 1. Bytes were valid at insertion
`POST /repository/draftfiles_ajax.php?action=list` returns, for the freshly imported file:

```json
{
  "filename": "The Adventures of Sherlock Holmes",
  "filepath": "/",
  "size": 47554,
  "mimetype": "Image (JPEG)",
  "image_width": 334,
  "image_height": 500,
  "url":          "https://moodle-playground.com/draftfile.php/5/user/draft/166589623/The%20Adventures%20of%20Sherlock%20Holmes",
  "realthumbnail":"https://moodle-playground.com/draftfile.php/5/user/draft/166589623/The%20Adventures%20of%20Sherlock%20Holmes?preview=thumb&oid=1778695080",
  "thumbnail":    "https://moodle-playground.com/theme/image.php/boost/core/1778694690/f/image"
}
```

`image_width` / `image_height` are only populated when `create_file_from_pathname()` successfully decodes the bytes via GD at insertion time. So the temp file *was* a complete, decodable JPEG. tcpOverFetch truncation, gzip-as-jpeg, and similar download-chain corruption theories are ruled out.

### 2. Hash-by-area succeeds, hash-by-`pathnamehash` fails
`draftfiles_ajax.php?action=list` finds the row via `get_area_files($context, 'user', 'draft', $itemid)` and returns it (proves DB row exists). But `draftfile.php` calls:

```php
$fs->get_file($context->id, 'user', 'draft', $itemid, '/', $filename)
```

which computes `sha1("/$contextid/$component/$filearea/$itemid$filepath$filename")` and looks up by `pathnamehash`. **That lookup returns `false`.** Stack trace (from `?debug=1`):

```
Error code: filenotfound
* line 2165 of /lib/filelib.php: core\exception\moodle_exception thrown
* line 81 of /draftfile.php: call to send_file_not_found()
```

### 3. Filename variants tested — all 404
We probed the scoped URL with several plausible filenames; every one returns 404:

| URL | status |
|---|---|
| `…/draftfile.php/5/user/draft/166589623/The%20Adventures%20of%20Sherlock%20Holmes`                       | 404 |
| `…/draftfile.php/…/The%20Adventures%20of%20Sherlock%20Holmes?oid=1778695080` (PR #91 fallback path)       | 404 |
| `…/draftfile.php/…/The%20Adventures%20of%20Sherlock%20Holmes.jpg`                                         | 404 |
| `…/draftfile.php/…/2ddb5c26bfb02963136e4b142ea397dae192c7ab.jpg` (original Omeka hash)                    | 404 |
| `…/pluginfile.php/5/user/private/0/The%20Adventures%20of%20Sherlock%20Holmes` (after Save Changes)        | 404 |

Filename in the DB is pure ASCII (`54686520416476656e7475726573…` — verified via raw byte dump).

### 4. SW routing is fine
Console shows `[playground] PHP exited 1 for /draftfile.php/…` — i.e. the SW *did* route the bare-origin URL to the runtime via referrer resolution and PHP did handle it. The 404 is from inside Moodle, not from a missing route. So this is **not** an architectural URL issue.

## Hypothesis

The `pathnamehash` column in the inserted `mdl_files` row is the SHA1 of an input string that differs from what Moodle's `clean_param(PARAM_FILE)` of the URL filename produces. Most likely causes (in order of probability):

1. The filename Moodle wrote (via the plugin's `get_file()` → `move_to_filepool()` chain) was different from the title shown in the picker (e.g. Moodle saved with `.jpg` extension or `clean_filename()` applied a normalization, but the picker's emitted URL doesn't reflect that). The `image_width` lookup works because it uses the area-query, not the hash-query.
2. A subtle `clean_param(PARAM_PATH)` round-trip on `filepath` introduces a normalization (e.g. trailing slash, encoding, multibyte) that differs from what the picker emits.
3. The repository download path under `@php-wasm` runs in a slightly different `core_text` config (e.g. ICU/iconv handling for Unicode normalization) than how the URL filename is later cleaned.

## Suggested next probe

Add a one-off PHP dump (admin script or `mtrace` in the plugin) that, immediately after the move-to-filepool, prints:

- The exact bytes of `filename` and `filepath` as stored in `mdl_files`
- The stored `pathnamehash`
- The expected `pathnamehash` recomputed from the SAME inputs

If those match, the lookup-time `clean_param` is normalizing differently. If they don't match, the plugin → framework chain stored a filename that doesn't round-trip through the URL.

## Note about PR #91

[PR #91](https://github.com/ateeducacion/moodle-playground/pull/91) (SW retry without `?preview=` on 404 for `draftfile.php` / `pluginfile.php`) is **harmless** here — its fallback request also 404s — and remains useful as a defence against the unrelated GD-thumbnail-generation failure mode it was originally designed for. Don't revert it; just understand it doesn't fix this issue.

**Comment by erseco** [comment]
Update from live testing on the pr-93 preview deploy:

- **URL downloader works** with the HEAD-body fix from #93 applied — confirmed by erseco.
- **`repository_omeka` still 404s** on `draftfile.php` with the same `filenotfound`, even though `draftfiles_ajax.php?action=list` returns the file with `size: 47554`, `image_width: 334`, `image_height: 500`.

So #93 (HEAD bug, now also filed upstream as WordPress/wordpress-playground#3632) fixed **a real, separate bug** that was visible in URL-downloader downloads — but it is **not** the cause of the omeka thumbnail 404s. The omeka-specific case remains unsolved and the `pathnamehash` mismatch hypothesis in this issue still stands.

Next probe (already shipped): a playground-gated `/repository/omeka/debug_filestorage.php` endpoint on branch [`debug/playground-filestorage`](https://github.com/ateeducacion/moodle-repository_omeka/tree/debug/playground-filestorage). Use this blueprint URL to load it:

```
https://pr-93--moodle-playground.netlify.app/?blueprint-url=https://raw.githubusercontent.com/ateeducacion/moodle-repository_omeka/refs/heads/debug/playground-filestorage/blueprint.json
```

Then visit `/repository/omeka/debug_filestorage.php?itemid=<draft-itemid>` (logged in) — the response dumps the actual stored `pathnamehash` / `contenthash` / filename byte hex / blob-on-disk existence vs recomputed hashes. That should pinpoint whether the bug is in Moodle's `clean_param(PARAM_FILE)`, `create_file_from_pathname()`, or somewhere else.
